drivers: wifi: nrf7002: Add support for multiple virtual interfaces

hanan619 opened the pull request:
**Description:**
The nRF7002 firmware supports two virtual interfaces (VIFs) that can operate in different modes (e.g., AP and STA). However, the existing Zephyr driver only utilizes a single VIF, preventing full multi-interface support.

This PR extends the nRF7002 driver to support multiple VIFs by making the following modifications:

- VIF Context Handling: The driver already contains an array of vif_ctx_zep, but only the first element was being used. Now, a second Ethernet device is registered using vif_ctx_zep[1], enabling multi-VIF operation.
- Tracking Interfaces: Introduced vif_ctx_cnt to keep track of active interfaces and manage their state effectively.
- FMAC Initialization: Ensured that FMAC (Firmware MAC) is initialized only once, avoiding redundant initializations when multiple VIFs are present.

**Testing:**
Verified that two virtual interfaces can be registered and operate simultaneously.
Tested AP mode on one VIF and STA mode on another VIF concurrently.
Ensured that UMAC command responses are correctly matched to their respective VIFs.
**Dependencies:**
This PR depends on the corresponding hostap PR, which introduces multiple control channels in WPA supplicant to handle separate interfaces.

krish2718 commented:
> @krish2718 Duplicate PR of #86618

Thanks, hopefully the tests will go through now.

MaochenWang1 commented on modules/hostap/src/wpa_cli.c:
```diff
@@ -21,8 +28,20 @@ static int cmd_wpa_cli(const struct shell *sh,
 			  size_t argc,
 			  const char *argv[])
 {
+	struct net_if *iface = net_if_get_first_wifi();
```
My concern is, if there are 2 virtual interfaces, then how to run `wpa_cli` cmd on the second interface, since you get the `iface` by `net_if_get_first_wifi`?

hanan619 replied:
A separate pr can be created to specify interface index in wpa cli.
This change is added to make the wpa cli compatible with new wpa supplicant changes as ctrl_conn is now associated with wpa_s object

krish2718 replied:
Fixed this.

krish2718 commented:
I did some testing and here's my feedback. BTW, how are you testing this PR? Are you testing on a nRF7002DK? Have you tested a working STA + AP? I see that the right after flashing the second VIF fails to be initialized?

```
uart:~$ 
[00:00:00.465,606] <err> wifi_nrf: nrf_wifi_if_start_zep: rpu_ctx_zep is NULL
*** Booting Zephyr OS build v4.1.0-6207-gf9f1f564fb93 ***
[00:00:00.484,436] <inf> wifi_supplicant: wpa_supplicant initialized
```

Earlier feedback is not implemented: https://github.com/zephyrproject-rtos/zephyr/pull/86618#issuecomment-2877378592

1. Missing fix: https://github.com/krish2718/zephyr/commit/73b149e9cbd01efdb223b56c9531c6aa60ba6648#diff-762ac3b8c00bc2d577de9dfcb9ce4d514d08233c991077269227fa889886929b -> Causes 2nd VIF to fail to initailize
2. Missing memory updates to hostap and socketpair (FYI, https://github.com/zephyrproject-rtos/zephyr/pull/90317 removes the socketpair support, so, might want to drop that change once its merged)

New issues

1. Fix Max VIF check in nrf_wifi (needs a manifest update as well)
2. Use the newly introduced: `WIFI_USAGE_MODE_STA_AP` (left a review comment too)

See attached patch with all fixes. (renamed as .patch isn't allowed in GH)
[mvif_working.txt](https://github.com/user-attachments/files/20842740/mvif_working.txt)


Build with recovery disabled (as I mentioned earlier, this doesn't work): `west build -p  -d samples/net/wifi/shell/build_70dk_dual -b nrf7002dk/nrf5340/cpuapp samples/net/wifi/shell/ -- -DCONFIG_LOG_MODE_IMMEDIATE=y -DCONFIG_WIFI_USAGE_MODE_STA_AP=y -DCONFIG_NRF_WIFI_RPU_RECOVERY=n`

```
uart:~$ wifi connect <...>
# Start AP on the same channel as STA
uart:~$ wifi ap enable -i 2 -s MVIF -c 2

uart:~$ wifi status -i 2
Status: successful
==================
State: COMPLETED
Interface Mode: ACCESS POINT
Link Mode: UNKNOWN
SSID: MVIF
BSSID: <SNIP..>
Band: 2.4GHz
Channel: 2
Security:  OPEN
MFP: Disable
Beacon Interval: 100
DTIM: 2
TWT: Not supported
Current PHY TX rate (Mbps) : 0
uart:~$ wifi status -i 1
Status: successful
==================
State: COMPLETED
Interface Mode: STATION
Link Mode: WIFI 6 (802.11ax/HE)
SSID: OpenWrt
BSSID: <SNIP..>
Band: 2.4GHz
Channel: 2
Security:  WPA2-PSK
MFP: Optional
RSSI: -74
Beacon Interval: 100
```

krish2718 committed:
drivers: wifi: nrf7002: Add support for multiple virtual interfaces (VIFs)

Description:
The nRF7002 firmware supports two virtual interfaces (VIFs) that can
operate in different modes (e.g., AP and STA). However, the existing
Zephyr driver only utilizes a single VIF, preventing full
multi-interface support.

This commit extends the nRF7002 driver to support multiple VIFs by
making the following modifications:

* The driver already contains an array of vif_ctx_zep, but only the
first item was being used. Now, a second Ethernet device is registered
using vif_ctx_zep[1], enabling multi-VIF operation.
* Introduced vif_ctx_cnt to keep track of active interfaces and manage
their state effectively.
* Ensured that FMAC (Firmware MAC) is initialized only once, avoiding
redundant initializations when multiple VIFs are present.
* The UMAC control commands previously did not associate responses with
the issuing VIF. A queue is now introduced to track the originating VIF
for each command and correctly route the response event to the
corresponding interface.

Signed-off-by: Hanan Arshad <hananarshad619@gmail.com>
Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
manifest: hostap: Pull hostap changes required for multiple VIF

This commit links the hostap PR required for the nrf7002 driver.
west.yml is updated

Signed-off-by: Hanan Arshad <hananarshad619@gmail.com>
Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
boards: nrf70: Add second virtual interface

Add second virtual interface by default to the DTS.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
modules: hostap: Update WPA supplicant to use per-VIF control channel

Update WPA supplicant functions to pass the control channel (socket) as a
parameter instead of relying on a global socket. This change aligns with
the PR 80 modifications in hostap repo and ensures that each Virtual
Interface (VIF) uses its dedicated control channel for communication.

Signed-off-by: Hanan Arshad <hananarshad619@gmail.com>
Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
wifi: treewide: Adjust defaults for dual mode usage

In case STA + AP mode is enabled, then adjust the defaults to accomodate
the second interface.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
modules: hostap: Adjust memories to handle dual mode

To handle the dual mode usage (STA + AP) increase the heap and stack
size.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
drivers: nrf_wifi: Fix scan crash for 2nd VIF

Move the VIF initialization to top to do it for both VIFs, this fixes a
crash when scanning on the 2nd VIF.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
modules: hostap: Enable AP mode if usage includes AP

If the usage mode has AP, then automatically enable AP mode.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
manifest: nrf_wifi: Pull fix for max VIF check

Fixes second interface bring up failure.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
modules: hostap: Add interface arg to cli commands

Now that we support multiple VIFs, need to add argument for the
interface for any command.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
modules: hostap: Use per-interface control connection

For hostapd also use per-interface control connection like we do for
supplicant.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>
